### PR TITLE
 table: handling the boundary value of the timestamp type's default value

### LINKED ISSUE #9513: Improper use of time zone when checking for default values ​in adding a column
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t(a int);
ALTER TABLE t ADD COLUMN ts timestamp DEFAULT '1970-01-01 08:00:01';
```

2. What did you expect to see?
```

mysql> select now();
+---------------------+
| now()               |
+---------------------+
| 2019-02-28 21:09:30 |
+---------------------+
1 row in set (0.00 sec)

mysql> create table t(a int);
Query OK, 0 rows affected (0.02 sec)

mysql> ALTER TABLE t ADD COLUMN ts timestamp DEFAULT '1970-01-01 08:00:01';
Query OK, 0 rows affected (0.03 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

3. What did you see instead?
```
tidb> select now();
+---------------------+
| now()               |
+---------------------+
| 2019-02-28 21:09:20 |
+---------------------+
1 row in set (0.00 sec)

tidb> create table t(a int);
Query OK, 0 rows affected (0.01 sec)

tidb>  ALTER TABLE t ADD COLUMN ts timestamp DEFAULT '1970-01-01 08:00:01';
ERROR 1067 (42000): Invalid default value for 'ts'
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v3.0.0-beta-132-gddbf60f6d
Git Commit Hash: ddbf60f6d013aadaba29676ccc47765a592b8841
Git Branch: master

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Fix #9513 
Before this PR:
```
tidb> select now();
+---------------------+
| now()               |
+---------------------+
| 2019-02-28 21:09:20 |
+---------------------+
1 row in set (0.00 sec)

tidb> create table t(a int);
Query OK, 0 rows affected (0.01 sec)

tidb>  ALTER TABLE t ADD COLUMN ts timestamp DEFAULT '1970-01-01 08:00:01';
ERROR 1067 (42000): Invalid default value for 'ts'
```
After this PR, we can successfully execute `ALTER TABLE t ADD COLUMN ts timestamp DEFAULT '1970-01-01 08:00:01';`.
### What is changed and how it works?
Using the correct time zone when checking the default value.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @crazycs520 @jackysp @tiancaiamao

**Comment by zimulala** [comment]
PTAL @crazycs520 @jackysp @tiancaiamao

File: table/column.go [hunk 1]
```diff
@@ -360,27 +360,36 @@ func getColDefaultValue(ctx sessionctx.Context, col *model.ColumnInfo, defaultVa
...
 	// Check and get timestamp/datetime default value.
-	value, err := expression.GetTimeValue(ctx, defaultVal, col.Tp, col.Decimal)
-	if err != nil {
-		return types.Datum{}, errGetDefaultFailed.GenWithStack("Field '%s' get default value fail - %s",
-			col.Name, errors.Trace(err))
-	}
-	// If column is timestamp, and default value is not current_timestamp, should convert the default value to the current session time zone.
+	sc := ctx.GetSessionVars().StmtCtx
+	var needChangeTimeZone bool
+	originalTZ := sc.TimeZone
```
Comment (crazycs520) [inline table/column.go]: move to after line #371.

File: table/column.go [hunk 2]
```diff
@@ -360,27 +360,36 @@ func getColDefaultValue(ctx sessionctx.Context, col *model.ColumnInfo, defaultVa
...
+			defer func() { sc.TimeZone = originalTZ }()
+		}
+	}
+	value, err := expression.GetTimeValue(ctx, defaultVal, col.Tp, col.Decimal)
+	if err != nil {
+		return types.Datum{}, errGetDefaultFailed.GenWithStack("Field '%s' get default value fail - %s",
+			col.Name, errors.Trace(err))
+	}
+	// If the column's default value is not ZeroDatetimeStr nor CurrentTimestamp, should convert the default value to the current session time zone.
+	if needChangeTimeZone {
```
Comment (xiekeyi98) [inline table/column.go]: Do we need more test to test this branch?

Comment (zimulala) [reply table/column.go]: @xiekeyi98  There have been many tests covering this situation before.

**Comment by zimulala** [comment]
PTAL @crazycs520

File: table/column.go (hunk 1 above)
Comment (zimulala) [reply table/column.go]: Done.

File: table/column.go [hunk 3]
```diff
@@ -360,27 +360,36 @@ func getColDefaultValue(ctx sessionctx.Context, col *model.ColumnInfo, defaultVa
...
 			// For col.Version = 0, the timezone information of default value is already lost, so use the system timezone as the default value timezone.
-			defaultTimeZone := timeutil.SystemLocation()
+			sc.TimeZone = timeutil.SystemLocation()
 			if col.Version >= model.ColumnInfoVersion1 {
-				defaultTimeZone = time.UTC
-			}
-			err = t.ConvertTimeZone(defaultTimeZone, ctx.GetSessionVars().Location())
-			if err != nil {
-				return value, errors.Trace(err)
+				sc.TimeZone = time.UTC
```
Comment (winkyao) [inline table/column.go]: Why do we need to change sc.TimeZone?

Comment (zimulala) [reply table/column.go]: Because when `col.Version >= ColumnInfoVersion1`(#9115) is true, the default value uses the UTC as the timezone.

Comment (winkyao) [reply table/column.go]: Please add some comments here, explain sc.TimeZone is used by expression.GetTimeValue, so this the reason why we change the sc.TimeZone.

Comment (winkyao) [reply table/column.go]: Maybe we should add a new parameter to pass the timezone to GetTimeValue? I affaid the defer will infect the performance.

Comment (XuHuaiyu) [reply table/column.go]: It seems that getTimeValue(ctx, xxx) will use sc.

Comment (zimulala) [reply table/column.go]: But the check function deep nested in `GetTimeValueis`, and these functions used a lot of places.  Besides, this change is a bit strange, there is already timezone in `sc`, we add a time zone parameter

**Comment by zimulala** [comment]
PTAL @winkyao

**Comment by zimulala** [comment]
PTAL @winkyao

**Comment by zimulala** [comment]
PTAL @winkyao

**Pull request merged**